syntax = "proto3";

package sensor_stream;

// Pose information
message Pose {
    double x = 1;
    double y = 2;
    double z = 3;
    double qw = 4;  // quaternion w
    double qx = 5;  // quaternion x
    double qy = 6;  // quaternion y
    double qz = 7;  // quaternion z
    uint64 timestamp_us = 8;
}

// Image frame
message ImageFrame {
    bytes data = 1;  // Compressed JPEG or raw bytes
    uint32 width = 2;
    uint32 height = 3;
    string encoding = 4;  // "rgb8", "bgr8", "jpeg", etc.
    uint64 timestamp_us = 5;
}

// Depth frame
message DepthFrame {
    bytes data = 1;  // Raw depth data or compressed
    uint32 width = 2;
    uint32 height = 3;
    string encoding = 4;  // "16UC1", "32FC1", etc.
    float depth_scale = 5;  // meters per unit
    uint64 timestamp_us = 6;
}

// Combined sensor data
message SensorData {
    Pose pose = 1;
    ImageFrame color_image = 2;
    DepthFrame depth_image = 3;
    uint64 frame_id = 4;

    // Raw bytes containing point coordinates
    bytes points_data = 5;
    uint32 num_points = 6;  // Number of points
}

// Command from server to drone
message DroneCommand {
    enum CommandType {
        NONE = 0;
        ARM = 1;
        TAKEOFF = 3;
        LAND = 4;
        FORWARD = 5;
        LEFT = 6;
        RIGHT = 7;
    }
    
    CommandType command = 1;
    
    // Optional parameters
    float x = 2;
    float y = 3;
    float z = 4;
    float velocity = 5;
}

// Response from drone
message CommandResponse {
    bool success = 1;
    string message = 2;
}

service SensorStreamService {
    // Bidirectional streaming
    rpc StreamSensorData(stream SensorData) returns (stream DroneCommand);
    
    // Separate command channel
    rpc SendCommand(DroneCommand) returns (CommandResponse);
}

